js/src/forum/utils/DiscussionControls.js

```diff
@@ -227,7 +228,13 @@ export default {
         app.history.back();
       }
 
-      return this.delete().then(() => app.discussions.removeDiscussion(this));
+      return this.delete().then(() => {
+        app.discussions.removeDiscussion(this);
+
+        if (app.current.matches(DiscussionsUserPage)) {
```

Comment:
Code like this (basically an `instanceof` check) always sets off alarm bells for me. `DiscussionControls` is a generic component, and you are coupling it to one of its current concrete usages.

You are also subtly changing behavior here, if I am not mistaken. The old code returned a promise that resolved to the result of `removeDiscussion` - now there's no more return value. That points to the proposed solution, though.

Please try to find a way to implement this with `deleteAction().then(...)` directly inside the `DiscussionsUserPage` component.

Reply:
The problem here is that deleteAction is never called from the DiscussionsUserPage; that page isn't even aware of it. In fact, all that page does is instantiate a DiscussionList and a DiscussionListState. The preferable solution would be some way to get the discussion list state instance to the discussion controls, but that also comes with it's own issues.